File: src/main/java/at/hrechny/predictionsbot/database/entity/GeneratedIdEntity.java
package at.hrechny.predictionsbot.database.entity;

import io.micronaut.core.annotation.Introspected;

import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import java.util.UUID;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.Hibernate;
import org.hibernate.annotations.GenericGenerator;

@Introspected(accessKind = Introspected.AccessKind.FIELD)
@Getter
@Setter
@MappedSuperclass
public abstract class GeneratedIdEntity {

  @io.micronaut.data.annotation.Id
  @Id
  @GeneratedValue(generator = "UUID")
  @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
  private UUID id;

  public UUID getId() {
    return id;
  }

  public void setId(UUID id) {
    this.id = id;
  }

  @Override
  public final boolean equals(Object object) {
    if (this == object) {
      return true;
    }
    if (object == null || Hibernate.getClass(this) != Hibernate.getClass(object)) {
      return false;
    }
    var that = (GeneratedIdEntity) object;
    return id != null && id.equals(that.id);
  }

  @Override
  public final int hashCode() {
    return Hibernate.getClass(this).hashCode();
  }

}

File: src/main/kotlin/at/hrechny/predictionsbot/database/entity/GeneratedIdEntity.kt
package at.hrechny.predictionsbot.database.entity

import io.micronaut.core.annotation.Introspected
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import java.util.UUID
import org.hibernate.Hibernate
import org.hibernate.annotations.UuidGenerator

@Introspected(accessKind = [Introspected.AccessKind.FIELD])
@MappedSuperclass
abstract class GeneratedIdEntity {
    @field:io.micronaut.data.annotation.Id
    @field:Id
    @field:GeneratedValue
    @field:UuidGenerator
    var id: UUID? = null

    final override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) {
            return false
        }

        other as GeneratedIdEntity
        return id != null && id == other.id
    }

    final override fun hashCode(): Int = Hibernate.getClass(this).hashCode()
}
